policies/codingstyle.txt

```diff
@@ -289,12 +289,69 @@ options.
 
 In source files, separate functions with one blank line.
 
+                Chapter 6.1: Function Arguments
+
+Most API functions in OpenSSL revolve around a specific type (e.g. SSL_CTX,
+EVP_PKEY, X509 etc). When adding new API functions relevant to a given type, the
+following argument ordering should be used:
+
+    rettype TYPE_get_something(TYPE *,[args])
+    rettype TYPE_do_something(TYPE *,[args])
+    TYPE *TYPE_new([args])
```

Comment:
This doesn't quite translate to the patterns we currently have.  I believe this is more complete / less ambiguous.

```
Some more complex complex APIs have a TYPE and TYPE_CTX pair, where TYPE usually is the set of function
pointers to the implementation, while TYPE_CTX is the collection of attributes to perform the intended operation.
For that case, the pattern should also include:

    TYPE_CTX *TYPE_CTX_new([args])
    rettype TYPE_CTX_get_something(TYPE_CTX *,[args])
    rettype TYPE_CTX_set_something(TYPE_CTX *,[args])
    rettype TYPE_do_something(TYPE_CTX *,[args])
```

Something like that.  To be noted is that for the complex case, `TYPE_do_something()` gets a changed first argument type, the rest are additions.  I don't expect the prose to be worded exactly like that, feel free to rephrase that to better fit the rest of the text.

Do we actually make APIs that don't have a TYPE / TYPE_CTX type of pair?  Do note that some APIs have very different names; for BIOs, that pair is BIO_METHOD / BIO (which is frankly a clearer naming for the same thing)

Reply:
This works for me - except to say shouldn't this *also* be ok as an *option*:

rettype TYPE_CTX_do_something(TYPE_CTX *,[args])

For example, in libssl there are some functions that really do operate on the SSL_CTX and aren't just simple getters/setters, e.g. `SSL_CTX_flush_sessions`

Reply:
I'm turning a blind eye to libssl, it seems to have patterns that work according to a (slightly? categorically?) different ruleset...

So then the question is, what to choose as a *guideline*?  In my view, some if not much of the SSL API divert from what we've discussed, so I'd rather see that as an exception to the guidelines.

Reply:
I've added the suggested text (with a few minor tweaks)

Reply:
> I've added the suggested text (with a few minor tweaks)

It seems like you didn't push those changes yet as a fixup?

Reply:
Oh, sorry, forget my comment. It's right below this box 😊 .

Reply:
> In my view, some if not much of the SSL API divert from what we've discussed, so I'd rather see that as an exception to the guidelines.

And the SSL API is the API that the majority of the users actually interact with so this is the pattern that the majority of the users are aware of. That you see it as not the same as some of the patterns in the Crypto API doesn't change the fact that it is the pattern visible to more users. Turning a _blind eye_ to the experiences and APIs that the majority of our users happen to use isn't the approach that I think makes sense here.